temporalio/worker/_workflow_instance.py

```diff
@@ -414,6 +419,100 @@ def _apply_cancel_workflow(
             # this cancellation to the next iteration of the event loop.
             self.call_soon(self._primary_task.cancel)
 
+    def _apply_do_update(
+        self, job: temporalio.bridge.proto.workflow_activation.DoUpdate
+    ):
+        # Run the validator & handler in a task. Everything, including looking up the update definition, needs to be
+        # inside the task, since the update may not be defined until after we have started the workflow - for example
+        # if an update is in the first WFT & is also registered dynamically at the top of workflow code.
+        async def run_update() -> None:
+            command = self._add_command()
+            command.update_response.protocol_instance_id = job.protocol_instance_id
+            try:
+                defn = self._updates.get(job.name) or self._updates.get(None)
+                if not defn:
+                    known_updates = sorted([k for k in self._updates.keys() if k])
+                    raise RuntimeError(
+                        f"Update handler for '{job.name}' expected but not found, and there is no dynamic handler. "
+                        f"known updates: [{' '.join(known_updates)}]"
+                    )
+                args = self._process_handler_args(
+                    job.name,
+                    job.input,
+                    defn.name,
+                    defn.arg_types,
+                    defn.dynamic_vararg,
+                )
+                handler_input = HandleUpdateInput(
+                    id=job.id,
+                    update=job.name,
+                    args=args,
+                    headers=job.headers,
+                )
+
+                # Always run the validator interceptor, which will only actually run a validator if one is defined.
```

Comment:
Subject to a discussion today (or soon) as we don't seem to have nailed this down yet

Reply:
Ok, after discussion, we have decided to only run this interceptor (and decode args for it) when there is a known validator on the definition

Reply:
Fixed